lib/puppet/provider/jail/pyiocage.rb

```diff
@@ -120,112 +121,126 @@ def set_property(property, value)
     iocage(['set', "#{property}=#{value}", resource[:name]])
   end
 
-  def state=(value)
-    @property_flush[:state] = value
+  def read_only(_value)
+    raise PuppetError, 'This property is read-only!'
   end
 
+  alias_method :jid=, :read_only
+
   def boot=(value)
     @property_flush[:boot] = value
   end
 
-  def ip4_addr=(value)
-    @property_flush[:ip4_addr] = value
+  def state=(value)
+    @property_flush[:state] = value
   end
 
-  def ip6_addr=(value)
-    @property_flush[:ip6_addr] = value
+  def type=(value)
+    @property_flush[:type] = value
+  end
+
+  def release=(value)
+    @property_flush[:release] = value
   end
 
-  def hostname=(value)
-    @property_flush[:hostname] = value
+  def ip4_addr=(value)
+    @property_flush[:ip4_addr] = value
   end
 
-  def jail_zfs=(value)
-    @property_flush[:jail_zfs] = value
+  def ip6_addr=(value)
+    @property_flush[:ip6_addr] = value
   end
 
-  def jail_zfs_dataset=(value)
-    @property_flush[:jail_zfs_dataset] = value
+  def template=(value)
+    @property_flush[:template] = value
   end
 
   def pkglist=(value)
     @property_flush[:pkglist] = value
   end
 
-  def wrap_create(jensure = resource[:ensure])
-    frel = Facter.value(:os)['release']['full'].gsub(%r{-p\d+$}, '')
-
-    template = resource[:template] ? "--template=#{resource[:template]}" : nil
-    release = resource[:release] ? "--release=#{resource[:release]}" : "--release=#{frel}"
-    from = template.nil? ? release : template
-
-    create_template = jensure == :template ? 'template=yes' : nil
+  def wrap_destroy
+    iocage(['destroy', '--force', resource[:name]])
+  end
 
-    unless resource[:pkglist].empty?
-      pkgfile = Tempfile.new('puppet-iocage-pkglist.json')
-      pkgfile.write({ pkgs: resource[:pkglist] }.to_json)
-      pkgfile.close
-      pkglist = "--pkglist=#{pkgfile.path}"
-    end
-    iocage(['create', '--force', from, pkglist, create_template, "tag=#{resource[:name]}"].compact)
+  # returns Optional[Tempfile] to the pkglist's contents
+  # users of this function should take care that it's deleted!
+  def create_pkglist(pkglist)
+    return nil if pkglist.nil? || pkglist.empty?
+    pkgfile = Tempfile.new('puppet-iocage.pkglist')
+    pkgfile.write({ pkgs: pkglist }.to_json)
+    pkgfile.close
+    pkgfile
   end
 
-  def wrap_destroy
-    iocage(['stop', resource[:name]])
-    iocage(['destroy', '--force', resource[:name]])
+  def wrap_create(options, props)
+    iocage('create', options, "--name #{resource[:name]}", props)
   end
 
-  def update
+  def rebuild(options, props)
     wrap_destroy
-    wrap_create
+    wrap_create(options, props)
   end
 
   def flush
+    options = []
+    props = []
     if @property_flush
```

Comment:
`@property_flush` would be an empty hash if none of the setter methods for the provider properties and parameters were called to set the relevant keys.  Those setter methods are only called in the event that something in the resource differs from reality, but I do see that you're settings a few items in the `@property_flush` hash.